R/utils.R

```diff
@@ -10,26 +10,107 @@
 #'
 #' @param db_data The REDCap database output defined by
 #' \code{REDCapR::redcap_read_oneshot()$data}
-#' @param var the unquoted name of the field containing event and arm
-#' identifiers
+#' @param var The unquoted name of the field containing event and arm
+#' identifiers. Default \code{NULL}.
 #'
 #' @importFrom dplyr mutate
+#' @importFrom rlang enexpr
 #'
 #' @keywords internal
 
 add_partial_keys <- function(db_data,
-                             var) {
-  pattern <- "^(\\w+?)_arm_(\\d)$"
+                             var = NULL) {
+
+  if (!is.null(enexpr(var))) {
+    pattern <- "^(\\w+?)_arm_(\\d)$"
+
+    db_data <- db_data %>%
+      mutate(
+        redcap_event = sub(pattern, "\\1", {{ var }}),
+        redcap_arm   = as.integer(sub(pattern, "\\2", {{ var }}))
+      )
+  }
 
   db_data <- db_data %>%
-    mutate(
-      redcap_event = sub(pattern, "\\1", {{ var }}),
-      redcap_arm   = as.integer(sub(pattern, "\\2", {{ var }}))
-    )
+    create_repeat_instance_vars()
 
   db_data
 }
 
+#' @title
+#' Utility function to convert redcap repeat instance columns into appropriate
+#' form and event columns
+#'
+#' @details
+#' The output of a standard REDCap export with repeating forms and/or events
+#' makes use of `redcap_repeat_instance` in combination with
+#' `redcap_repeat_instrument` and whether or not data exists in both. Instead,
+#' rename and separate `redcap_repeat_instance` into `redcap_form_instance` and
+#' `redcap_event_instance`.
+#'
+#' @param db_data The REDCap database output generated by
+#' \code{REDCapR::redcap_read_oneshot()$data}
+#'
+#' @return
+#' A dataframe.
+#'
+#' @importFrom dplyr rename_with mutate case_when pull relocate
+#' @importFrom tidyselect matches
+#'
+#' @keywords internal
+
+create_repeat_instance_vars <- function(db_data) {
+
+  out <- db_data
+
+  # Detect if repeat forms exist
+
+  has_repeat_forms <- "redcap_repeat_instance" %in% names(db_data)
+
+  if (has_repeat_forms) {
+    out <- db_data %>%
+      rename_with(~"redcap_form_instance",
+                  .cols = matches("redcap_repeat_instance"))
+  }
+
+  # Detect if repeat events exist
+  #   Determined by non-NA vals in new "redcap_form_instance" alongside
+  #   NA vals in "redcap_repeat_instrument"
+  #   `has_repeat_forms` will always be TRUE for events to exist
+  if (has_repeat_forms) {
+    has_repeat_events <- out %>%
+      mutate(
+        repeat_events_check = case_when(is.na(redcap_repeat_instrument) & !is.na(redcap_form_instance) ~ TRUE,
+                                        TRUE ~ FALSE)
+      ) %>%
+      pull(.data$repeat_events_check) %>% #nolint: object_usage_linter
+      unique() %>%
+      any()
```

Comment:
is.na(redcap_repeat_instrument) & !is.na(redcap_form_instance)
```suggestion
    has_repeat_events <- any(
        is.na(out$redcap_repeat_instrument) & !is.na(out$redcap_form_instance)
    )
```

Reply:
Much simpler 😅

Reply:
base R ftw 😂